pkg/app_gtwy/server/aether-app-gtwy-impl.go

```diff
@@ -0,0 +1,684 @@
+		if _, ok := imsiMap[imsi]; ok {
+			ip := string(imsiMap[imsi].Metric["ip"])
+			newDev.Ip = &ip
+			attached := imsiMap[imsi].Value.String()
+			newDev.Attached = &attached
+		}
+
+		listDeviceResponse = append(listDeviceResponse, newDev)
+	}
+
+	log.Infof("GetAppGtwyDevices")
+	return ctx.JSON(http.StatusOK, listDeviceResponse)
+}
+
+// GetDevice /appgtwy/v1/{target}/enterprises/{enterprise-id}/sites/{site-id}/devices/{device-id}
+func (i *AppGtwy) GetDevice(ctx echo.Context, target externalRef0.Target, enterpriseID string, siteID string, deviceId string) error {
+
+	gnmiCtx, cancel := utils.NewGnmiContext(ctx, i.GnmiTimeout)
+	defer cancel()
+
+	site, err := i.gnmiGetEnterprisesEnterpriseSite(gnmiCtx, "/aether/v2.0.0/{target}/enterprises/enterprise/{enterprise-id}/site/{site-id}", target, enterpriseID, siteID)
+	if err != nil {
+		return utils.ConvertGrpcError(err)
+	}
+
+	si, err := i.AnalyticsClient.Query("subscriber_info")
+	if err != nil {
+		return err
+	}
+
+	simMap := mapSim(site)
+	deviceGroupMap := mapDeviceToDeviceGroups(site)
+	imsiMap := mapIMSI(si)
+
+	var device types.AppGtwyDevice
+	for _, d := range *site.Device {
+		if d.DeviceId != deviceId {
+			continue
+		}
+		s := simMap[*d.SimCard]
+		deviceID := &d.DeviceId
+		device = types.AppGtwyDevice{
+			DeviceId:     deviceID,
+			DisplayName:  d.DisplayName,
+			Description:  d.Description,
+			Imei:         d.Imei,
+			SimIccid:     s.Iccid,
+			DeviceGroups: deviceGroupMap[deviceId],
+		}
+
+		imsi := strconv.Itoa(int(*s.Imsi))
+		if _, ok := imsiMap[imsi]; ok {
+			ip := string(imsiMap[imsi].Metric["ip"])
+			device.Ip = &ip
+			attached := imsiMap[imsi].Value.String()
+			device.Attached = &attached
+		}
+	}
+
+	log.Infof("GetDevice")
+	return ctx.JSON(http.StatusOK, device)
+}
+
+// gnmiGetEnterprisesEnterpriseSite returns an instance of Enterprises_Enterprise_Site.
+func (i *AppGtwy) gnmiGetEnterprisesEnterpriseSite(ctx context.Context,
+	openAPIPath string, target externalRef0.Target, args ...string) (*externalRef0.EnterprisesEnterpriseSite, error) {
+
+	gnmiGet, err := utils.NewGnmiGetRequest(openAPIPath, string(target), args...)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("gnmiGetRequest %s", gnmiGet.String())
+	gnmiVal, err := utils.GetResponseUpdate(i.GnmiClient.Get(ctx, gnmiGet))
+	if err != nil {
+		return nil, err
+	}
+	if gnmiVal == nil {
+		return nil, nil
+	}
+	gnmiJSONVal, ok := gnmiVal.Value.(*gnmi.TypedValue_JsonVal)
+	if !ok {
+		return nil, fmt.Errorf("unexpected type of reply from server %v", gnmiVal.Value)
+	}
+
+	log.Debugf("gNMI Json %s", string(gnmiJSONVal.JsonVal))
+	var gnmiResponse externalRef1.Device
+	if err = externalRef1.Unmarshal(gnmiJSONVal.JsonVal, &gnmiResponse); err != nil {
+		return nil, fmt.Errorf("error unmarshalling gnmiResponse %v", err)
+	}
+	mpd := ModelPluginDevice{
+		device: gnmiResponse,
+	}
+
+	return mpd.toEnterprisesEnterpriseSite(args...)
+}
+
+// ModelPluginDevice - a wrapper for the model plugin
```

Comment:
you could split this Model Plugin related stuff out in to a separate file to keep the concepts cleaner

Reply:
Agreed!